exercises/nondet_monad_defun/descr.md

```diff
@@ -0,0 +1,530 @@
+is worth noting that `xs` is itself a function, so the elements of the
+sequence `xs` need not be explicitly computed until `xs` is applied.
+
+Sequences are closely related to *iterators* in object-oriented languages,
+such as C++ and Java. Yet, sequences are much simpler than iterators, for
+two reasons:
+
+* they involve no mutable state;
+
+* they are just as easy to construct and to use as ordinary lists.
+
+The functions `nil`, `cons`, and `singleton` are constructor functions.
+
+The functions `map`, `concat`, `flatten` are analogues for sequences
+of the standard list functions `List.map`, `(@)`, and
+`List.flatten`.
+
+The function `Seq.take` truncates a sequence at a certain length:
+`Seq.take n xs` is a sequence that begins like `xs` but has at most
+`n` elements.
+
+The function `Seq.head` demands the first element of a sequence. If
+the sequence begins with an element `x`, then `Some x` is returned;
+otherwise, `None` is returned. This forces enough computation to
+take place so as to be able to produce the first element of the
+sequence.
+
+The functions `Seq.of_list` and `Seq.to_list` convert between lists
+and sequences, both ways. One must keep in mind that applying
+`Seq.to_list` to a sequence `xs` causes all of its elements to be
+demanded: that is, it forces all of the suspended computations to
+take place. In particular, if `xs` is an infinite sequence, then
+`Seq.to_list xs` does not terminate.
+
+## Data Structures for Continuations and Computations
+
+In this exercise, we implement an abstract machine that manipulates three data
+structures: a **computation**, a **failure continuation**, and a **success
+continuation**. The idea is, in general, the computation indicates what to do
+next; in the case where the computation is `fail`, the failure continuation
+indicates what to do next; in the case where the computation is `return x`,
+the success continuation indicates what to do next.
+
+Each of these three data structures is described by a generalized algebraic
+data type. We now give the definitions of these data types. (These definitions
+can in fact be *deduced* from the solution of the
+<a href="" onclick="top.location='/exercises/nondet_monad_cont/';">companion exercise</a>,
+by defunctionalization, but we leave this story for another time.)
+
+The type `'a m` of **computations** is defined as follows:
+
+```
+type 'a m =
+| MDelay:               (unit -> 'a m) -> 'a m
+| MReturn:                          'a -> 'a m
+| MBind:           'a m * ('a -> 'b m) -> 'b m
+| MFail:                                  'a m
+| MChoose:                 'a m * 'a m -> 'a m
+| MReflect: ('a * 'a m) option failure -> 'a m
+```
+
+The data constructors `MDelay`, `MReturn`, `MBind`, `MFail`, and `MChoose`
+correspond directly to the constructor functions `delay`, `return`, `>>=`,
+`fail`, and `choose`. In other words, a computation is represented internally
+by its syntax.
```

Comment:
Can we say that this is an instance of a Free Monad?

Reply:
Maybe; I am not sure what the term "free monad" means exactly. Do you think that I should add this comment? Would be it technically accurate, and would it be helpful?

Reply:
François' filters of truth make me realize now that it is indeed not useful. :-)